Enable WASM AOT option in build and update MonoAOTCompiler

akoeplinger opened the pull request:
Allows creating AOT images for WASM with the AOT compiler and consuming them in the runtime.

Use `make build AOT=1` in `src/mono/netcore/sample/wasm/console` to try it out.

akoeplinger committed:
Enable WASM AOT option in build and update MonoAOTCompiler

vargaz commented on src/mono/mono.proj:
```diff
@@ -186,7 +186,7 @@
     </PropertyGroup>
     <ItemGroup Condition="'$(TargetsBrowser)' == 'true'">
-      <_MonoCMakeArgs Include="-DENABLE_MINIMAL=ssa,com,jit,reflection_emit_save,portability,assembly_remapping,attach,verifier,appdomains,shadowcopy,security,sgen_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,logging,remoting,shared_perfcounters,sgen_debug_helpers,sgen_binary_protocol,soft_debug,interpreter,cleanup,mdb,gac,threads,eventpipe,aot,interpreter,qcalls$(_MonoMinimal) ...
+      <_MonoCMakeArgs Include="-DENABLE_MINIMAL=ssa,com,jit,reflection_emit_save,portability,assembly_remapping,attach,verifier,appdomains,shadowcopy,security,sgen_marksweep_conc,sgen_split_nursery,sgen_gc_bridge,sgen_toggleref,logging,remoting,shared_perfcounters,sgen_debug_helpers,sgen_binary_protocol,soft_debug,interpreter,cleanup,mdb,gac,threads,eventpipe,qcalls$(_MonoMinimal)"/>
```
The 'interpreter' option will disable the interpreter as part of the runtime, and compile it into a separate lib using the -DENABLE_INTERP_LIB=1 line.

akoeplinger replied:
it's still there, I just removed the duplicated instance of `interpreter` in the line

vargaz commented on src/mono/netcore/sample/Android/Makefile:
```diff
@@ -2,10 +2,7 @@ MONO_CONFIG=Release
-
-runtimepack:
-	../../../../.././build.sh Mono+Libs -os Android -arch $(MONO_ARCH) -c $(MONO_CONFIG)
+all: run
```
This was a helper target used during development.

akoeplinger replied:
it's building quite a lot of stuff, I thought everyone is using the `make runtime` target in src/mono/wasm/Makefile now :)

if you think it's useful I can bring this back and call that target instead.

vargaz commented on src/mono/netcore/sample/iOS/Makefile:
```diff
@@ -10,9 +10,6 @@ appbuilder:
 
-runtimepack:
-	../../../../.././build.sh Mono+Libs -os iOS -arch $(MONO_ARCH) -c $(MONO_CONFIG)
-
```
Ditto.

vargaz commented on src/mono/CMakeLists.txt:
```diff
@@ -290,6 +290,9 @@ elseif(TARGET_SYSTEM_NAME STREQUAL "Emscripten")
   endif()
+  if(NOT HOST_WASM)
+    set(GC_SUSPEND "default")
+  endif()
```
Why is this needed ? It should only be hit for the cross compiler.

akoeplinger replied:
yes this is for the cross compiler, I was hitting some issue otherwise.

vargaz replied:
Wasn't this fixed by:
https://github.com/dotnet/runtime/pull/44093
?

akoeplinger replied:
ah yeah, I'll check.

akoeplinger replied:
removed it.

akoeplinger committed:
PR feedback

akoeplinger committed:
Unify name of the msbuild property to enable AOT with iOS and disable a few runtime features on wasm sample

akoeplinger committed:
Merge remote-tracking branch 'upstream/master' into wasm-aot

akoeplinger committed:
Update after shared frameworks sdk PR

akoeplinger committed:
Simplify WasmSample.csproj

akoeplinger committed:
Fix regression in aot-compiler.c regarding SPC->corlib assembly name

In https://github.com/dotnet/runtime/commit/035ebeab160b9a281e5fbb3fa0c237f244c81b7a we added a simplified prefix for symbol names in System.Private.CoreLib.
However this shouldn't happen for the actual assembly name otherwise we aren't able to find the AOT module linked statically into the binary.

akoeplinger commented:
Test failure is due to https://github.com/dotnet/runtime/issues/44626.
